destination.go

```diff
@@ -1,75 +1,190 @@
+	d.conn, err = ampqDial(d.config.URL, d.tlsConfig)
+	if err != nil {
+		return fmt.Errorf("failed to dial: %w", err)
+	}
+	sdk.Logger(ctx).Debug().Str("url", d.config.URL).Msg("connected to RabbitMQ")
+
+	d.ch, err = d.conn.Channel()
+	if err != nil {
+		return fmt.Errorf("failed to open channel: %w", err)
+	}
+	sdk.Logger(ctx).Debug().Msgf("opened channel")
+
+	_, err = d.ch.QueueDeclare(
+		d.config.Queue.Name,
+		d.config.Queue.Durable,
+		d.config.Queue.AutoDelete,
+		d.config.Queue.Exclusive,
+		d.config.Queue.NoWait,
+		nil)
+	if err != nil {
+		return fmt.Errorf("failed to declare queue: %w", err)
+	}
+	sdk.Logger(ctx).Debug().Any("queueConfig", d.config.Queue).Msgf("declared queue")
+
+	if d.config.Exchange.Name != "" {
+		err = d.ch.ExchangeDeclare(
+			d.config.Exchange.Name,
+			d.config.Exchange.Type,
+			d.config.Exchange.Durable,
+			d.config.Exchange.AutoDelete,
+			d.config.Exchange.Internal,
+			d.config.Exchange.NoWait,
+			nil,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to declare exchange: %w", err)
+		}
+		sdk.Logger(ctx).Debug().Any("exchange config", d.config.Exchange).Msgf("declared exchange")
+
+		err = d.ch.QueueBind(d.config.Queue.Name, d.config.RoutingKey, d.config.Exchange.Name, false, nil)
+		if err != nil {
+			return fmt.Errorf("failed to bind queue to exchange: %w", err)
+		}
+		sdk.Logger(ctx).Debug().Msgf(
+			"bound queue %s to exchange %s with routing key %s",
+			d.config.Queue.Name, d.config.Exchange.Name, d.config.RoutingKey,
+		)
+	}
+
 	return nil
 }
 
 func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
-	// Write writes len(r) records from r to the destination right away without
-	// caching. It should return the number of records written from r
-	// (0 <= n <= len(r)) and any error encountered that caused the write to
-	// stop early. Write must return a non-nil error if it returns n < len(r).
-	return 0, nil
+	for _, record := range records {
+		msgID := string(record.Position)
+		msg := amqp091.Publishing{
+			ContentType:     d.config.Delivery.ContentType,
+			ContentEncoding: d.config.Delivery.ContentEncoding,
+			DeliveryMode:    d.config.Delivery.DeliveryMode,
+			Priority:        d.config.Delivery.Priority,
+			CorrelationId:   d.config.Delivery.CorrelationID,
+			ReplyTo:         d.config.Delivery.ReplyTo,
+
+			MessageId: msgID,
+			Type:      d.config.Delivery.MessageTypeName,
+			UserId:    d.config.Delivery.UserID,
+			AppId:     d.config.Delivery.AppID,
+			Body:      record.Bytes(),
+
+			Expiration: d.config.Delivery.Expiration,
+		}
+
+		if createdAt, err := record.Metadata.GetCreatedAt(); err != nil {
+			msg.Timestamp = createdAt
+		}
+
+		err := d.ch.PublishWithContext(
+			ctx,
+			d.config.Exchange.Name,
+			d.config.RoutingKey,
+			d.config.Delivery.Mandatory,
+			d.config.Delivery.Immediate,
+			msg,
+		)
+		if err != nil {
+			return 0, fmt.Errorf("failed to publish: %w", err)
```

Comment:
the int number returned is the number of records that were already written, so in this loop, it could be more than 0, so we should return `i` which is the index of the record we're currently writing

Reply:
Out of curiosity, does conduit actually use the total written records when encountering a write error? or you mention this just so that the destination connector better aligns with the `io.Writer` interface?